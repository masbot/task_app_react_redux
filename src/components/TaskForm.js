import React, { Component } from 'react';
import { Form, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap';

class TaskForm extends Component {
    constructor() {
        super();
        this.state = {
            task: ''
        }
    }

    render() {
        return (
            <div>
                <Form>
                    <FormGroup>
                        <FormControl 
                            placeholder="NEW TASK"
                            className="input-task" 
                            onChange={event => this.setState({task: event.target.value})}
                        />
                    </FormGroup>
                </Form>
                <Button className="btn btn-add-newtask" onClick={() => this.props.addNewTask(this.state.task)}>Submit</Button>
            </div>
        )
    }
}

export default TaskForm;