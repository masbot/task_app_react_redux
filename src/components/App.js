import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { addNewTask, fetchTasklist, saveTasklist, removeTask } from '../actions/tasklist';
import { clearError } from '../actions/errorhandler';
import TaskForm from './TaskForm';
import Task from './Task';

export class App extends Component {
    constructor() {
        super();
        this.state = {
            taskformDisplay: false
        }
    }

    componentDidMount() {
        this.props.fetchTasklist();
    }

    addTask = () => {
        this.setState({ taskformDisplay: true });
    }

    addNewTask = (task) => {
        this.props.addNewTask(task);
        this.setState({taskformDisplay: false });
    }

    remove = (id) => {
        this.props.removeTask(id);
    }

    save = () => {
        this.props.saveTasklist(this.props.tasks);
    }

    clear = () => {
        this.props.clearError();
    }

    createMessage = (errortype) => {
        let msg;
        let type;

        if(errortype === 'fetch-error'){
            msg = 'Please Reload The Page';
            type = 'error';
        }else if(errortype === 'save-error'){
            msg = 'Please Try Again';
            type = 'error';
        }else if(errortype === 'success'){
            msg = 'Your List Has Been Successfully Saved!';
            type = 'success';
        }
        msg = <div className={`alert ${type}`}>
                <span>{msg}</span>
                <i className="fa fa-times-circle" onClick={this.clear} aria-hidden="true"/>
             </div>;

        return msg;
    }

    render() {
        let taskform;
        const tasklist = [];
        const errorhandler = this.props.errorhandler;
        let msg;

        if( errorhandler) {
            msg = this.createMessage(errorhandler);
        }

        if(this.state.taskformDisplay){
            taskform = <TaskForm addNewTask={this.addNewTask}/>;
        }

        if(Object.keys(this.props.tasks).length != 0){
            let tasks = this.props.tasks;
            for(let key in tasks){
                tasklist.push(<Task key={key} id={key} task={tasks[key]} remove={this.remove}/>)
            }
        }

        return (
            <div className="container">
                <header>
                    <div className="h_1"><span className='title'>Tasks</span></div>
                    <div className="h_2">
                        <Button className="btn btn-add-task" onClick={this.addTask}>Add Task</Button>
                        <Button className="btn btn-save-tasklist" onClick={this.save}>Save</Button>
                    </div>
                </header>
                <section>
                    <div className="error-msg">
                        {msg}
                    </div>
                    <div className="task-form">
                        { taskform }
                    </div>
                    <div className="task-list">
                        { tasklist.reverse() }
                    </div>
                </section>
            </div>
        )
    }
}

export default connect(state => state, {fetchTasklist, saveTasklist, addNewTask, removeTask, clearError})(App);