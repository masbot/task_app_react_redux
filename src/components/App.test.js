import React from 'react';
import { mount, shallow } from 'enzyme';
import { App } from './App';
import TaskForm from './TaskForm';

describe('App', () => {
    const mockFetchtasklist = jest.fn();
    let props = { tasks: {"1": "TASK"}};
    let app = shallow(<App {...props}/>);
    const task = "BUY MILK";
    
    it('renders correctly', () => {
        expect(app).toMatchSnapshot();
    });
    
    it('TaskForm does not appear', () => {
        expect(app.find('TaskForm').exists()).toBe(false);
    });

    describe('clicking the `add task`', () => {
        beforeEach(() => {
            app.find('.btn-add-task').simulate('click');
        });
        
        it('TaskForm appears', () => {
            expect(app.find('TaskForm').exists()).toBe(true);
        });
    });

    describe('when mounted', () => {
        beforeEach(() => {
            let props = { tasks: {"1": "TASK"}};
            props.fetchTasklist = mockFetchtasklist;
            app = mount(<App {...props}/>);
        });

        it('dispatches the `fetchTasklist()` method it receives from props', () => {
            expect(mockFetchtasklist).toHaveBeenCalled();
        });
    });

    describe('when showing error and success messages', () => {
        beforeEach(() => {
            let props = {tasks: {"1": "TASK"}, errorhandler: 'fetch-error'};
            app = shallow(<App  {...props}/>)
        })

        it('shows error message when fetching for data fails', () => {
            expect(app.find('.error').exists()).toBe(true);
        });
    });

});