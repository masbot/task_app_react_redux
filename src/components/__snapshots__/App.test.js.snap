// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    tasks={
        Object {
            "1": "TASK",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header>
          <div
                    className="h_1"
          >
                    <span
                              className="title"
                    >
                              Tasks
                    </span>
          </div>
          <div
                    className="h_2"
          >
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-add-task"
                              disabled={false}
                              onClick={[Function]}
                    >
                              Add Task
                    </Button>
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-save-tasklist"
                              disabled={false}
                              onClick={[Function]}
                    >
                              Save
                    </Button>
          </div>
</header>,
        <section>
          <div
                    className="error-msg"
          />
          <div
                    className="task-form"
          />
          <div
                    className="task-list"
          >
                    <Task
                              id="1"
                              remove={[Function]}
                              task="TASK"
                    />
          </div>
</section>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="h_1"
>
              <span
                            className="title"
              >
                            Tasks
              </span>
</div>,
            <div
              className="h_2"
>
              <Button
                            active={false}
                            block={false}
                            bsClass="btn"
                            bsStyle="default"
                            className="btn btn-add-task"
                            disabled={false}
                            onClick={[Function]}
              >
                            Add Task
              </Button>
              <Button
                            active={false}
                            block={false}
                            bsClass="btn"
                            bsStyle="default"
                            className="btn btn-save-tasklist"
                            disabled={false}
                            onClick={[Function]}
              >
                            Save
              </Button>
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span
                className="title"
>
                Tasks
</span>,
              "className": "h_1",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Tasks",
                "className": "title",
              },
              "ref": null,
              "rendered": "Tasks",
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="btn btn-add-task"
                  disabled={false}
                  onClick={[Function]}
>
                  Add Task
</Button>,
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="btn btn-save-tasklist"
                  disabled={false}
                  onClick={[Function]}
>
                  Save
</Button>,
              ],
              "className": "h_2",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": false,
                  "block": false,
                  "bsClass": "btn",
                  "bsStyle": "default",
                  "children": "Add Task",
                  "className": "btn btn-add-task",
                  "disabled": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Add Task",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": false,
                  "block": false,
                  "bsClass": "btn",
                  "bsStyle": "default",
                  "children": "Save",
                  "className": "btn btn-save-tasklist",
                  "disabled": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Save",
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="error-msg"
/>,
            <div
              className="task-form"
/>,
            <div
              className="task-list"
>
              <Task
                            id="1"
                            remove={[Function]}
                            task="TASK"
              />
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "error-msg",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "task-form",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Task
                  id="1"
                  remove={[Function]}
                  task="TASK"
/>,
              ],
              "className": "task-list",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "id": "1",
                  "remove": [Function],
                  "task": "TASK",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "section",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <div
                        className="h_1"
            >
                        <span
                                    className="title"
                        >
                                    Tasks
                        </span>
            </div>
            <div
                        className="h_2"
            >
                        <Button
                                    active={false}
                                    block={false}
                                    bsClass="btn"
                                    bsStyle="default"
                                    className="btn btn-add-task"
                                    disabled={false}
                                    onClick={[Function]}
                        >
                                    Add Task
                        </Button>
                        <Button
                                    active={false}
                                    block={false}
                                    bsClass="btn"
                                    bsStyle="default"
                                    className="btn btn-save-tasklist"
                                    disabled={false}
                                    onClick={[Function]}
                        >
                                    Save
                        </Button>
            </div>
</header>,
          <section>
            <div
                        className="error-msg"
            />
            <div
                        className="task-form"
            />
            <div
                        className="task-list"
            >
                        <Task
                                    id="1"
                                    remove={[Function]}
                                    task="TASK"
                        />
            </div>
</section>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="h_1"
>
                <span
                                className="title"
                >
                                Tasks
                </span>
</div>,
              <div
                className="h_2"
>
                <Button
                                active={false}
                                block={false}
                                bsClass="btn"
                                bsStyle="default"
                                className="btn btn-add-task"
                                disabled={false}
                                onClick={[Function]}
                >
                                Add Task
                </Button>
                <Button
                                active={false}
                                block={false}
                                bsClass="btn"
                                bsStyle="default"
                                className="btn btn-save-tasklist"
                                disabled={false}
                                onClick={[Function]}
                >
                                Save
                </Button>
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span
                  className="title"
>
                  Tasks
</span>,
                "className": "h_1",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Tasks",
                  "className": "title",
                },
                "ref": null,
                "rendered": "Tasks",
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="btn btn-add-task"
                    disabled={false}
                    onClick={[Function]}
>
                    Add Task
</Button>,
                  <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="btn btn-save-tasklist"
                    disabled={false}
                    onClick={[Function]}
>
                    Save
</Button>,
                ],
                "className": "h_2",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "active": false,
                    "block": false,
                    "bsClass": "btn",
                    "bsStyle": "default",
                    "children": "Add Task",
                    "className": "btn btn-add-task",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Add Task",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "active": false,
                    "block": false,
                    "bsClass": "btn",
                    "bsStyle": "default",
                    "children": "Save",
                    "className": "btn btn-save-tasklist",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Save",
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="error-msg"
/>,
              <div
                className="task-form"
/>,
              <div
                className="task-list"
>
                <Task
                                id="1"
                                remove={[Function]}
                                task="TASK"
                />
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "error-msg",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "task-form",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Task
                    id="1"
                    remove={[Function]}
                    task="TASK"
/>,
                ],
                "className": "task-list",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "class",
                  "props": Object {
                    "id": "1",
                    "remove": [Function],
                    "task": "TASK",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "section",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
